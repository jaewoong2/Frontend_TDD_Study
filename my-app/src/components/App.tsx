import styled from 'styled-components';
import { Route, Routes } from 'react-router-dom';
import { List } from './Pages/List';
import Form from './Pages/Form';
import { Detail } from './Pages/Detail';
import PageHeader from './blocks/PageHeader';

function App() {
  return (
    <Container>
      <Contents>
        <PageHeader />
        <Routes>
          <Route path="/" element={<List />} />
          <Route path="/add" element={<Form />} />
          <Route path="/detail/:id" element={<Detail />} />
          <Route path="*" element={<div className="error">Error!</div>} />
        </Routes>
      </Contents>
    </Container>
  );
}

export default App;

const Container = styled.div`
  width: 100%;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Contents = styled.div`
  max-width: 1024px;
  display: flex;
  flex-direction: column;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  position: relative;
  align-content: center;

  .error {
    width: 400px;
    height: 450px;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 1.3em;
  }
`;
