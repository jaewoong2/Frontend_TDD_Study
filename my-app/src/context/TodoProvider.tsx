import React, { createContext, useEffect, useMemo, useState } from 'react';

const initalAction = {
  addList: () => (todo: string) => {},
  removeList: (index: number) => {},
};

export const TodoActionsContext = createContext(initalAction);
export const TodoValueContext = createContext<string[]>([]);

type Props = {
  children?: React.ReactNode;
};

export function TodoProvider({ children }: Props) {
  const { addStroage, getStorage, removeStroage } = utils;
  const [todoLists, setTodoLists] = useState<string[]>(getStorage());

  const actions = useMemo(
    () => ({
      addList: () => (todo: string) => {
        if (todo.trim() !== '') {
          setTodoLists((prev) => [...prev, todo]);
          addStroage(todo);
        }
      },
      removeList: (index: number) =>
        setTodoLists((prev) => {
          removeStroage(index);
          return [...prev].filter((_, i) => i !== index);
        }),
    }),
    [],
  );

  return (
    <TodoActionsContext.Provider value={actions}>
      <TodoValueContext.Provider value={todoLists}>
        {children}
      </TodoValueContext.Provider>
    </TodoActionsContext.Provider>
  );
}

const utils = {
  getStorage: () => {
    const data = localStorage.getItem('todoItem');
    if (data !== null) {
      const todolists = JSON.parse(data);
      if (todolists) {
        return todolists;
      }
      return [];
    }
    return [];
  },

  addStroage: (item: string) => {
    const data = utils.getStorage();
    const lists = [...data, item];
    localStorage.setItem('todoItem', JSON.stringify(lists));
  },

  removeStroage: (index: number) => {
    const data = utils.getStorage();
    const lists = [...data].filter((_, i) => i !== index);
    localStorage.setItem('todoItem', JSON.stringify(lists));
  },
};
