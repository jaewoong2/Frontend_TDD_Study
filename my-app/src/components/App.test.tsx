import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { TodoProvider } from 'context/TodoProvider';
import 'jest-styled-components';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import App from './App';

function ProvidedApp() {
  return (
    <TodoProvider>
      <App />
    </TodoProvider>
  );
}

describe('<App />', () => {
  it('renders component correctly', () => {
    const { container } = renderWithRouter(<ProvidedApp />, { route: '/' });

    const todoList = screen.getByTestId('toDoList');
    expect(todoList).toBeInTheDocument();
    expect(todoList.firstChild).toBeNull();

    const header = screen.getByText('할 일 목록');
    expect(header).toBeInTheDocument();

    const label = screen.getByText('+');
    expect(label).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('All Test', () => {
    renderWithRouter(<ProvidedApp />, { route: '/' });

    const label = screen.getByText('+');
    expect(label).toBeInTheDocument();
    fireEvent.click(label);

    expect(screen.getByText('할 일 추가')).toBeInTheDocument();

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
    expect(input).toBeInTheDocument();
    fireEvent.change(input, { target: { value: 'study react' } });

    const addButton = screen.getByText('+ ADD');
    fireEvent.click(addButton);

    const goBack = screen.getByText('Back');
    fireEvent.click(goBack);

    expect(screen.getByText('할 일 목록')).toBeInTheDocument();

    const todoList = screen.getByTestId('toDoList');
    expect(todoList).toBeInTheDocument();

    const todoItem = screen.getByText('study react');
    expect(todoItem).toBeInTheDocument();

    const moveButton = screen.getByText('Move to Item');
    fireEvent.click(moveButton);

    expect(screen.getByText('할 일 상세')).toBeInTheDocument();

    const removeButton = screen.getByText('Delete');
    fireEvent.click(removeButton);
    fireEvent.click(screen.getByText('Back'));

    expect(screen.getByText('할 일 목록')).toBeInTheDocument();
    expect(screen.getByTestId('toDoList').firstChild).toBeNull();
  });
});

// needs for react-router-dom v6
const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
