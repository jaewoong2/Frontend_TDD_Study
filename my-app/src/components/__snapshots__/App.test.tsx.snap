// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders component correctly 1`] = `
.c2 {
  width: 100%;
  height: 40px;
  background-color: #9bceee;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 8px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 .box {
  width: 100%;
}

.c2 .center {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  font-size: 20px;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 40px;
  height: 40px;
  bottom: -30px;
  border-radius: 30%;
  cursor: pointer;
  background-color: #304ffe;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:hover {
  background-color: #1e40ff;
}

.c5:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c0 {
  width: 100%;
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  max-width: 1024px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

.c1 .error {
  width: 400px;
  height: 450px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.3em;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <nav
        class="c2"
      >
        <div
          class="box left"
        >
          <a
            href="/"
          >
            Back
          </a>
        </div>
        <div
          class="box center"
        >
          할 일 목록
        </div>
        <div
          class="box right"
        />
      </nav>
      <div
        class="c3"
      >
        <ul
          class="c4"
          data-testid="toDoList"
        />
        <a
          class="c5"
          href="/add"
        >
          +
        </a>
      </div>
    </div>
  </div>
</div>
`;
